bika/lims/browser/js/bika.lims.analysisrequest.add.js

```diff
@@ -80,40 +82,71 @@
+      $("body").on("click", "tr[fieldname=Composite] input[type='checkbox']", this.recalculate_records);
+      $("body").on("click", "tr[fieldname=InvoiceExclude] input[type='checkbox']", this.recalculate_records);
+      $("body").on("click", "tr[fieldname=Analyses] input[type='checkbox']", this.on_analysis_checkbox_click);
+      $("body").on("selected change", "tr[fieldname=Client] input[type='text']", this.on_client_changed);
+      $("body").on("selected change", "tr[fieldname=Contact] input[type='text']", this.on_contact_changed);
+      $("body").on("click", "tr[fieldname=ReportDryMatter] input[type='checkbox']", this.on_reportdrymatter_click);
+      $("body").on("change", "input.min", this.on_analysis_specification_changed);
+      $("body").on("change", "input.max", this.on_analysis_specification_changed);
+      $("body").on("change", "input.err", this.on_analysis_specification_changed);
+      $("body").on("click", ".service-lockbtn", this.on_analysis_lock_button_click);
+      $("body").on("click", ".service-infobtn", this.on_analysis_details_click);
+      $("body").on("selected change", "tr[fieldname=Sample] input[type='text']", this.on_sample_changed);
+      $("body").on("selected change", "tr[fieldname=SampleType] input[type='text']", this.on_sampletype_changed);
+      $("body").on("selected change", "tr[fieldname=Specification] input[type='text']", this.on_specification_changed);
+      $("body").on("selected change", "tr[fieldname=Template] input[type='text']", this.on_analysis_template_changed);
+      $("body").on("selected", "tr[fieldname=Profiles] input[type='text']", this.on_analysis_profile_selected);
+      $("body").on("click", "tr[fieldname=Profiles] img.deletebtn", this.on_analysis_profile_removed);
+      $("body").on("click", "img.copybutton", this.on_copy_button_click);
 
       /* internal events */
-      $(this).on("form:changed", this.recalculate_records);
-      $(this).on("data:updated", this.update_form);
-      $(this).on("data:updated", this.recalculate_prices);
-      $(this).on("data:updated", this.hide_all_service_info);
+      $(this).on("form:changed", this.debounce(this.recalculate_records, 500));
+      $(this).on("data:updated", this.debounce(this.recalculate_prices, 3000));
+      $(this).on("data:updated", this.debounce(this.update_form, 300));
+      $(this).on("data:updated", this.debounce(this.hide_all_service_info, 300));
       $(this).on("ajax:start", this.on_ajax_start);
       return $(this).on("ajax:end", this.on_ajax_end);
     };
 
+    AnalysisRequestAdd.prototype.debounce = function(func, threshold, execAsap) {
+
+      /*
+       * Debounce a function call
+       * See: https://coffeescript-cookbook.github.io/chapters/functions/debounce
+       */
+      var timeout;
+      timeout = null;
+      return function() {
+        var args, delayed, obj;
+        args = 1 <= arguments.length ? slice.call(arguments, 0) : [];
```

Comment:
Understand here that you allow the function to consider additional arguments not explicitly set as a parameter passed-in through the function signature, but haven't seen nowhere the assignment of `arguments`. I am missing something?

Reply:
Yes, the debounced functions do not handle the `event` argument that comes in automatically from `on`. But when you do a `console.log(arguments)` in `update_form`, you see that the arguments are passed through the same way as w/o the debounce wrapper.